### PR TITLE
feat(analyzer): implement useNodeImportProtocol rule

### PR DESCRIPTION
**Pull request by 2-NOW**
## Summary

Implement the [prefer-node-protocol](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-node-protocol.md): for node builtin modules, write `"node:\<builtin>"` instead of `"\<builtin>"` when import modules.

```js
// invalid
import path from "path"

// valid
import os from "node:os"
```

closes: #818

## Test Plan

Added related tests inside.

### CODE REVIEW

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            return None;
+        };
+
+        if !is_builtin_module(&module_name) {
+            return None;
+        }
+
+        Some(NodeImportProtocolState {
+            range: binding.range(),
+            module_name: module_name.to_string(),
```
Comment (ematipico) [inline crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs:76]: Since you can compute the module name from binding, it's best to return nothing - Som(()) - and compute it  inside the diagnostic function. Like this, we don't allocate a string

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                "Import from Node.js builtin module \""{module_name}"\" should use the \"node:\" protocol."
+            },
+        )
+        .note(markup! {
+            "Please change to \"node:"{module_name}"\"."
+        }))
+    }
+}
+
+const NODE_BUILTINS: &[&str] = &[
```
Comment (ematipico) [inline crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs:96]: Some advice:
- node contains also promises modules, like `fs/promises`. These should be taken into account
- If you sort alphabetically this array, you can use the binary search method instead of contains, which is faster
- you should take into account `require()` and `import()`
- there's another rule being developed - `noNodejsBuiltins` - which adds a similar array. Once your rule and the other are merged, we should have the same array in single place: src/globals/node.rs

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs (hunk 1 above)
Comment (2-NOW) [reply crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs]: Thanks! I've made the [changes](https://github.com/biomejs/biome/pull/1131/files#diff-3ccea7f725f28ab591e458af19749a00bd9419f029d8f155d6ce76dd52361772R68-R87) you suggested.

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs (hunk 2 above)
Comment (2-NOW) [reply crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs]: - I added more [node builtins](https://github.com/biomejs/biome/pull/1131/files#diff-7e43cb70df659129a98cbc96f2ab9f06f5936dd8988d30829b50bdf919c19540R71) and changed the location to `globals/node.rs` for other rules that will be added.
- I sorted the array alphabetically and changed contains to a [binary search](https://github.com/biomejs/biome/pull/1131/files#diff-3ccea7f725f28ab591e458af19749a00bd9419f029d8f155d6ce76dd52361772R89-R91).

However, I couldn't make it work for commonjs. I tried, but the JsModuleSource only worked for esm.
Can you give me a hint what I can try to make it work for cjs ?

Comment (ematipico) [reply crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs]: > However, I couldn't make it work for commonjs. I tried, but the JsModuleSource only worked for esm.
Can you give me a hint what I can try to make it work for cjs ?

In this case, you have to query multiple nodes. To do so, you can use the macro `declare_node_union!`. We have many rules that do so, you can take inspiration from them.  You'll have to query the node [`JsCallExpression`](https://biomejs.dev/playground/?code=cgBlAHEAdQBpAHIAZQAoACIAZgBzACIAKQA%3D), and check that `callee` is a `"require"` and it has one argument, which belongs to the node built-ins

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            binding.range(),
+            markup! {
+                "Import from Node.js builtin module \""{module_name}"\" should use the \"node:\" protocol."
+            },
+        )
+        .note(markup! {
+            "Please change to \"node:"{module_name}"\"."
+        }))
```
Comment (ematipico) [inline crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs:84]: I believe we can suggest a `code_action` to change it, but we can do it a different PR if you prefer

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs (hunk 2 above)
Comment (2-NOW) [reply crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs]: Thanks! I'll try it as you said.

File: crates/biome_js_analyze/src/globals/node.rs [hunk 4]
```diff
@@ -68,6 +68,63 @@ pub const NODE: [&str; 34] = [
...
+    "url",
+    "util",
+    "util/types",
+    "v8",
+    "vm",
+    "wasi",
+    "worker_threads",
+    "zlib",
+];
+
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/globals/node.rs:127]: We should add in `fn test_order` a test that ensures that NODE_BUILTIN is sorted.

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    fn diagnostic(ctx: &RuleContext<Self>, _: &Self::State) -> Option<RuleDiagnostic> {
+        let binding: &JsModuleSource = ctx.query();
+        let module_name = binding.inner_string_text().ok()?.to_string();
+
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            binding.range(),
+            markup! {
+                "Import from Node.js builtin module \""{module_name}"\" should use the \"node:\" protocol."
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs:76]: We could add emphasis:
```suggestion
                "Import from Node.js builtin module \""<Emphasis>{module_name}</Emphasis>"\" should use the \""<Emphasis>"node:"</Emphasis>"\" protocol."
```

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+impl Rule for UseNodeImportProtocol {
+    type Query = Ast<JsModuleSource>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let binding: &JsModuleSource = ctx.query();
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs:56]: I suggest removing the type annotation because it is trivially inferred and to rename `binding` to `source`.
```suggestion
        let source = ctx.query();
```

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        let Ok(module_name) = binding.inner_string_text() else {
+            return None;
+        };
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs:59]: You can simply use `.ok()?` to obtain the result :)
```suggestion
        let module_name = binding.inner_string_text().ok()?;
```

File: crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        if !is_builtin_module(&module_name) {
+            return None;
+        }
+
+        Some(())
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs:65]: Inthsi particular case, you could use `then_some`
```suggestion
        (!is_builtin_module(&module_name)).then_some(())
```

**Comment by anonrig** [comment]
This is really similar to my PR (https://github.com/biomejs/biome/pull/1121). Would you like to take over it, and add this to this PR?

**Comment by 2-NOW** [comment]
> This is really similar to my PR (#1121). Would you like to take over it, and add this to this PR?

Sure! As you said, the PRs you posted seem to be similar, so I'll take over the PRs and add them in

**Comment by ematipico** [comment]
I am going to merge this in a temporary branch and finish the works. Unfortunately I have permission issues when trying to push to this fork.
